text/0012-open-host-info-to-plugins.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: pluggable_host_info
+- Start Date: 2016-08-18
+- RFC PR:
+- Issue:
+
+# Summary
+[summary]: #summary
+
+Make the ENC output produced by Foreman extendable by plugins.
+
+# Motivation
+[motivation]: #motivation
+
+As part of the effort to remove puppet from core, I found that the ENC output
+represented by `Host#info` method contains elements from different parts of
+the system. To break this monolith, I want to introduce providers registry so
+the info output will be constructed by aggregation of the output from each
+provider.
+
+One of the most vital constructs in the core that should be taken into account
+in this redesign is the parameters framework represented by the `Classification`
+module. This module is responsible for dealing with key-value storage offered
+by the core and extended in modules like puppet. Basically it is responsible
+for defining which value is relevant to each one of selected keys, given a
+`Host` instance.
+
+Today I am aware of two more requirements that could be implemented with the
+help of the suggested framework:
+1. Chef plugin would like to extend the ENC by its unique information.
+2. OpenSCAP wants to enrich the info from puppet provider.
+
+# Detailed design
+[design]: #detailed-design
+
+## Registry
+We will start with a simple registry to store the list of registered providers.
+The registry will be a simple module with a property to hold the list of
+providers and a valid accessor method.
+
+``` ruby
+module HostInfo
+  class << self
+    def info_providers
+      @info_providers ||= []
+    end
+
+    def register_info_provider(provider)
+      info_providers << provider
+    end
+  end
+end
+```
+
+The registry should also be available through the plugin declaration, in order
+to ease the usage in the framework:
+
+``` ruby
+
+plugin :puppet do
+  provides_info PuppetHostInfoProvider
+end
+
+```
+
+## Provider
+All providers should inherit a base class that will define the semi-formal
+interface for interactions with the provider:
+
+``` ruby
+class HostInfo::Provider
+  def initialize(host)
+    @host = host
+  end
+
+  # Will return provider-specific portion of the info hash.
+  def host_info
+    throw 'Provider is not implemented'
+  end
+
+  protected
```

Comment:
To reduce the public interface to a minimum. The `:host` attribute is there for the implementers to be able to access the host. It's part of the class context, rather than a property that it manages.